### PR TITLE
More resilient blob handling in snapshot repositories

### PR DESCRIPTION
**Pull request by abeyad**
This PR encompasses work done to enhance the resiliency of snapshot repository blob handling.  In particular:
1. deleteBlob methods in the `Repository` interface now throw an IOException if the blob does not exist.
2. Due to some test scenarios, this enhancement caused test failures, because blob names were based on the snapshot name (instead of snapshot UUID), which led to non-unique (i.e. mutable) blobs. So the deleteBlob change had to be accompanied by naming repository blobs by their snapshot UUIDs instead of snapshot names, which was part of the plan in #18156 
3. To consolidate all blob naming changes to one PR in order to avoid incompatibility issues, we also added the notion of an index ID (separate from the index UUID) used to name indices in the repositories `indices` folder, in order to prevent invalid characters in blob names as described in #7540

This PR is a combination of two other PRs:
#18815
#19421

Relates #18156

### CODE REVIEW

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java [hunk 1]
```diff
@@ -116,6 +117,11 @@ private OutputStream createOutput(String blobName) throws IOException {
     @Override
     public void deleteBlob(String blobName) throws IOException {
         logger.trace("deleteBlob({})", blobName);
+
+        if (!blobExists(blobName)) {
+            throw new IOException("Blob [" + blobName + "] does not exist");
```
Comment (dadoonet) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java]: Should it be a `NoSuchFileException` here?

File: plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/MockAmazonS3.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.DigestInputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+class MockAmazonS3 extends AbstractAmazonS3 {
```
Comment (dadoonet) [inline plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/MockAmazonS3.java]: So great you wrote that class.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java (hunk 1 above)
Comment (abeyad) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java]: @dadoonet Great catch, thanks! I will fix it before merging

**Comment by abeyad** [comment]
This PR has passed testing on CI twice as well as locally for me, so I'm merging and will monitor the builds over the next few days.

**Comment by abeyad** [comment]
@gfyoung many thanks for your great work on getting this through!

**Pull request merged**